### PR TITLE
feat: add home view artists rail

### PR DESCRIPTION
**Pull request by MounirDhahri**
This PR resolves [ONYX-1167] 

### Description
This PR adds the artists generic rail to `Section`. This rail's goal is to render `ArtistsRailHomeViewSection`.

https://github.com/user-attachments/assets/f35cda56-9c0b-4083-9478-4ca6e4269723

### PR Checklist

- [x] I have tested my changes on **iOS** and **Android**.
- [x] I hid my changes behind a **[feature flag]**, or they don't need one.
- [x] I have included **screenshots** or **videos**, or I have not changed the UI.
- [x] I have added **tests**, or my changes don't require any.
- [x] I added an **[app state migration]**, or my changes do not require one.
- [x] I have documented any **follow-up work** that this PR will require, or it does not require any.
- [x] I have added a **changelog entry** below, or my changes do not require one.

### To the reviewers 👀

- [ ] I would like **at least one** of the reviewers to **run** this PR on the simulator or device.

<details><summary>Changelog updates</summary>

#### Cross-platform user-facing changes

-

#### iOS user-facing changes

-

#### Android user-facing changes

-

#### Dev changes

-

</details>

Need help with something? Have a look at our [docs], or get in touch with us.

[app state migration]: ../blob/main/docs/adding_state_migrations.md
[feature flag]: ../blob/main/docs/developing_a_feature.md
[docs]: ../blob/main/docs/README.md

[ONYX-1167]: https://artsyproduct.atlassian.net/browse/ONYX-1167?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by ArtsyOpenSource** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Warnings</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:warning:</td>
      <td>

  ❌ **An error occurred while validating your changelog, please make sure you provided a valid changelog.**
  </td>
    </tr>
  </tbody>
</table>



<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against a788a025da34303482c4d1941190a678bb8f4909
</p>

File: src/app/Components/Home/ArtistRails/ArtistCard.tsx [hunk 1]
```diff
@@ -16,7 +16,7 @@ interface ArtistCardProps {
   onPress?: () => void
 }
 
-const IMAGE_MAX_HEIGHT = 180
+export const IMAGE_MAX_HEIGHT = 180
```
Comment (olerichter00) [inline src/app/Components/Home/ArtistRails/ArtistCard.tsx:19]: Might be good to choose a name related to the artist rail if we export it. Could be something like:

```suggestion
export const ARTIST_RAIL_MAX_IMAGE_SIZE = 180
```

File: src/app/Components/Home/ArtistRails/RecommendedArtistsRail.tsx [hunk 2]
```diff
@@ -85,7 +85,7 @@ export const RecommendedArtistsRail: React.FC<RecommendedArtistsRailProps & Rail
       </Flex>
       <CardRailFlatList<ArtistCard_artist$data>
         listRef={listRef}
-        prefetchUrlExtractor={(item) => item?.href!}
+        prefetchUrlExtractor={(item) => item?.href || undefined}
```
Comment (olerichter00) [inline src/app/Components/Home/ArtistRails/RecommendedArtistsRail.tsx:88]: Instead of adding `|| undefined`, we can probably also accept `null` [here](https://github.com/artsy/eigen/blob/db2e0f09ed52ba8927fb0fa2b9a054a9fb3491a9/src/app/Components/PrefetchFlatList.tsx#L11).

File: src/app/Components/Home/ArtistRails/RecommendedArtistsRail.tsx [hunk 3]
```diff
@@ -180,7 +180,7 @@ export const tracks = {
   }),
 }
 
-const followOrUnfollowArtist = (followArtist: ArtistCard_artist$data) => {
+export const followOrUnfollowArtist = (followArtist: ArtistCard_artist$data) => {
```
Comment (olerichter00) [inline src/app/Components/Home/ArtistRails/RecommendedArtistsRail.tsx:183]:
```suggestion
const followOrUnfollowArtist = (followArtist: ArtistCard_artist$data) => {
```

File: src/app/Scenes/HomeView/Sections/ArtistsRailHomeViewSection.tsx [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+              )
+            }
+            return null
+          }}
+          renderItem={({ item: artist, index }) => {
+            return (
+              <View style={{ flexDirection: "row" }}>
+                <ArtistCardContainer
+                  artist={artist}
+                  // onFollow={() => followOrUnfollowArtist(artist)}
```
Comment (olerichter00) [inline src/app/Scenes/HomeView/Sections/ArtistsRailHomeViewSection.tsx:75]: 👀

File: src/app/Components/Home/ArtistRails/RecommendedArtistsRail.tsx (hunk 3 above)
Comment (olerichter00) [reply src/app/Components/Home/ArtistRails/RecommendedArtistsRail.tsx:183]: https://github.com/artsy/eigen/pull/10551/files#diff-26a95c958ab313802cb7c4d321bccd0d993c935cd97d0d32acae54b4ac9975a8R75

File: src/app/Scenes/HomeView/Sections/ArtistsRailHomeViewSection.tsx [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+        <CardRailFlatList<Artist>
+          data={artists}
+          keyExtractor={(artist) => artist.internalID}
+          // I noticed that sometimes FlatList seemed to get confused about where cards should be
+          // and making this explicit fixes that.
+          getItemLayout={(_data, index) => ({
+            index,
+            offset: index * (CARD_WIDTH + INTER_CARD_PADDING),
+            length: CARD_WIDTH + INTER_CARD_PADDING,
+          })}
```
Comment (olerichter00) [inline src/app/Scenes/HomeView/Sections/ArtistsRailHomeViewSection.tsx:53]: Is this something we could also do in `CardRailFlatList.tsx`? Looks like something we should apply everywhere and not here 🤔

File: src/app/Scenes/HomeView/Sections/ArtistsRailHomeViewSection.tsx [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+          onEndReached={onEndReached}
+          ListFooterComponent={() => {
+            if (!!hasNext && isLoadingNext) {
```
Comment (olerichter00) [inline src/app/Scenes/HomeView/Sections/ArtistsRailHomeViewSection.tsx:56]:
```suggestion
            if (hasNext && isLoadingNext) {
```

File: src/app/Scenes/HomeView/Sections/ArtistsRailHomeViewSection.tsx [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+                  width={IMAGE_MAX_HEIGHT / 2}
+                  height={IMAGE_MAX_HEIGHT}
+                  justifyContent="center"
+                  alignItems="center"
+                >
+                  <Spinner />
+                </Flex>
+              )
+            }
+            return null
```
Comment (olerichter00) [inline src/app/Scenes/HomeView/Sections/ArtistsRailHomeViewSection.tsx:68]:
```suggestion
            }

            return null
```

File: src/app/Scenes/HomeView/Sections/ArtistsRailHomeViewSection.tsx [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+          renderItem={({ item: artist, index }) => {
+            return (
+              <View style={{ flexDirection: "row" }}>
+                <ArtistCardContainer
+                  artist={artist}
+                  // onFollow={() => followOrUnfollowArtist(artist)}
+                />
+                {index === artists.length - 1 ? null : (
+                  <View style={{ width: INTER_CARD_PADDING }} />
+                )}
```
Comment (olerichter00) [inline src/app/Scenes/HomeView/Sections/ArtistsRailHomeViewSection.tsx:79]: Would this be accurate / equivalent? 

```suggestion
                {(index < artists.length - 1)  && <View style={{ width: INTER_CARD_PADDING }} />}

```

File: src/app/Scenes/HomeView/Sections/ArtistsRailHomeViewSection.tsx [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+
+type Artist = ExtractNodeType<ArtistsRailHomeViewSection_section$data["artistsConnection"]>
+export const ArtistsRailHomeViewSection: React.FC<ArtworksRailHomeViewSectionProps> = ({
+  section,
+}) => {
+  const { data, hasNext, loadNext, isLoadingNext } = usePaginationFragment(artistsFragment, section)
+
+  const title = data.component?.title
+
+  const onEndReached = useCallback(() => {
```
Comment (araujobarret) [inline src/app/Scenes/HomeView/Sections/ArtistsRailHomeViewSection.tsx:28]: don't think we need the `useCallback` here since we don't use `onEndReached` in any `useEffect`

File: src/app/Components/Home/ArtistRails/RecommendedArtistsRail.tsx (hunk 3 above)
Comment (araujobarret) [reply src/app/Components/Home/ArtistRails/RecommendedArtistsRail.tsx:183]: I think this shouldn't be here anymore, [idea] we could move it to: 
```
srr/app/Components/Home/mutations/followOrUnfollowArtist.ts
```

File: src/app/Components/Home/ArtistRails/ArtistCard.tsx (hunk 1 above)
Comment (MounirDhahri) [reply src/app/Components/Home/ArtistRails/ArtistCard.tsx:19]: I feel like it's kind of clear here since it's already insde the `ArtistCard` component. I will update it in the import thought to be something like
```typescript
import { IMAGE_MAX_HEIGHT as ARTIST_RAIL_MAX_IMAGE_SIZE } from ...
```

File: src/app/Components/Home/ArtistRails/RecommendedArtistsRail.tsx (hunk 3 above)
Comment (MounirDhahri) [reply src/app/Components/Home/ArtistRails/RecommendedArtistsRail.tsx:183]: we do actually have a helper for that in [here](https://github.com/artsy/eigen/blob/f993f231211b0797e2755ca00966522442fd9843/src/app/utils/mutations/useFollowArtist.ts#L3)
https://github.com/artsy/eigen/blob/f993f231211b0797e2755ca00966522442fd9843/src/app/utils/mutations/useFollowArtist.ts#L3
but I am avoiding making big changes here because we are not keeping any of the rails and mainly trying to hook the data as an example on how to consume the new HomeView API

**Review by MounirDhahri — COMMENTED** [review]
thanks for the early comments - I pushed this code last week to make sure it can make it to the Home View updates call before wrapping up the PR.

File: src/app/Scenes/HomeView/Sections/ArtistsRailHomeViewSection.tsx (hunk 5 above)
Comment (olerichter00) [reply src/app/Scenes/HomeView/Sections/ArtistsRailHomeViewSection.tsx]: I agree if this is a hack/fix specifically for only this component. But if it is not, I would suggest applying it in the `CardRailFlatList`. If we don't, wouldn't the next one using `CardRailFlatList` most likely forget to test for this bug and add the fix 🤔

File: src/app/Scenes/HomeView/Sections/ArtistsRailHomeViewSection.tsx [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+  return (
+    <Flex>
+      <Flex pl={2} pr={2}>
+        <SectionTitle title={title} />
+      </Flex>
+      <CardRailFlatList<Artist>
+        data={artists}
+        keyExtractor={(artist) => artist.internalID}
+        onEndReached={onEndReached}
+        ItemSeparatorComponent={() => <Spacer x={`${INTER_CARD_PADDING}px`} />}
```
Comment (araujobarret) [inline src/app/Scenes/HomeView/Sections/ArtistsRailHomeViewSection.tsx:49]: We shouldn't be using this `INTER_CARD_PADDING`(sorry was hard to follow up all the PRs once this was introduced) at all, we have our spacing defined in the DS level and we should use it instead of abstractions on top of it.
Another important thing is the value of this variable is not fine, it should be 10 or 20, we don't support 15.

File: src/app/Scenes/HomeView/Sections/ArtistsRailHomeViewSection.tsx [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+            return null
+          }}
+          renderItem={({ item: artist, index }) => {
+            return (
+              <View style={{ flexDirection: "row" }}>
+                <ArtistCardContainer
+                  artist={artist}
+                  // onFollow={() => followOrUnfollowArtist(artist)}
+                />
+                {index === artists.length - 1 ? null : (
```
[the remark this answers is not in the document]
Comment (araujobarret) [reply src/app/Scenes/HomeView/Sections/ArtistsRailHomeViewSection.tsx]: 😢

File: src/app/Scenes/HomeView/Sections/ArtistsRailHomeViewSection.tsx (hunk 10 above)
Comment (MounirDhahri) [reply src/app/Scenes/HomeView/Sections/ArtistsRailHomeViewSection.tsx:49]: that's a good point - I totally agree here.
I am trying to avoid making any changes to existing components because I don't have enough context on why it was made that way. This component is mainly a copy paste to recommended artists rails. Design is coming up with new rails soon

Comment (MounirDhahri) [reply src/app/Scenes/HomeView/Sections/ArtistsRailHomeViewSection.tsx:49]: I will update the value to `1` though and hope design doesn't complain about it 😄

Comment (araujobarret) [reply src/app/Scenes/HomeView/Sections/ArtistsRailHomeViewSection.tsx]: 🤞🤞🤞 
Usually, that's a smoke signal they've made the mockups inconsistent with our DS(using a spacing measure that doesn't exist)

Comment (araujobarret) [reply src/app/Scenes/HomeView/Sections/ArtistsRailHomeViewSection.tsx]: that's the problem of reusing the existing elements which I mentioned before, it snowballs these elements' debt even more

File: src/app/Components/Home/ArtistRails/ArtistCard.tsx (hunk 1 above)
Comment (olerichter00) [reply src/app/Components/Home/ArtistRails/ArtistCard.tsx:22]: I guess, it's ok like it is 👍 I'm just concerned about exporting things with a generic name because it could confuse editor's auto-import feature...

**Comment by MounirDhahri** [comment]
Let's merge this and move to hooks later - @olerichter00 thanks for the fixes you made.
